accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,362 @@
+
+  <packageSourceCredentials>
+    <ExamplePrivateRepo>
+      <add key="ToolName" value="Microsoft.CredentialProvider" />
+      <!-- The value here is the .NET tool name installed globally or in a specific tool-path -->
+    </ExamplePrivateRepo>
+    <!-- Add credentials for other package sources as needed -->
+  </packageSourceCredentials>
+</configuration>
+```
+
+**Advantages:**
+
+- This approach explicitly configures the intent to use the .NET Tool as a plugin for authentication in the NuGet.Config file itself.
+- Customers can install the plugins as global .NET Tools.
+
+**Disadvantages:**
+
+- Configuring the setting per feed can be cumbersome if multiple private feeds can use the same .NET tool for authentication.
+
+An alternative approach to address this disadvantage is to declare the plugins explicitly outside of the package source sections.
+This approach allows customers to specify the dependency only once, without configuring it per private feed as discussed above.
+
+```xml
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <DotNetToolPlugins>
+    <Authentication>
+      <plugin id="CredentialProvider.Microsoft" />
+      <plugin id="CredentialProvider.GitHub" />
+      <!-- Allows specifying multiple authentication plugins with additional details -->
+    </Authentication>
+  <!-- This format provides flexibility for future plugin types and detailed configurations -->
+  </DotNetToolPlugins>
+</configuration>
+```
+
+However, the NuGet Client plugins functionality is a global setting.
+This means that the plugins are discovered based on the platform and target framework, as discussed earlier, and there is currently no way to configure the dependent plugins via NuGet settings per repository.
+It could be a future possibility to provide users with an option to manage their plugins per repository instead of loading all the available plugins.
+However, given the limited number of plugin implementations currently available, this is not a problem that needs to be solved at this point in time, in my understanding.
+
+## Prior Art
+
+- The AWS team has developed a [.NET tool](https://docs.aws.amazon.com/codeartifact/latest/ug/nuget-cli.html#nuget-configure-cli) to simplify authentication with their private NuGet feeds.
+In their current implementation, they've added a subcommand, `codeartifact-creds install`.
+This command copies the credential provider to the NuGet plugins folder.
+They also provide an `uninstall` subcommand to remove the files from the NuGet plugin folders.
+However, due to limitations in the NuGet Client tooling, they've had to maintain plugins for both .NET Framework and .NET Core tooling. Additionally, they've had to provide commands to install and uninstall the NuGet plugins.
+
+- The `dotnet workload` command is separate and has its own set of sub-commands, including `install`, `uninstall`, and `list`.
+These sub-commands are wrappers for the corresponding `dotnet tool` sub-commands.
```

Comment:
Either we have different definitions of what a wrapper means, or no they're not.

To me, a first command that wraps a second command means that the first command invokes the second command to do its work. This implies that the functionality of the first command can wholly be replicated by using the second command directly. It's just that the first command provides a nicer interface.

Here's the MAUI workload package: https://nuget.info/packages/Microsoft.NET.Sdk.Maui.Manifest-6.0.400/6.0.553
When I look at the nuspec, I see it has `<packageType name="DotnetPlatform" />`.
Also the package contents are in a `data/` directory.

When I look at a dotnet tool package: https://nuget.info/packages/dotnet-suggest/1.1.415701
I see the nuspec has `<packageType name="DotnetTool" />`.
The package contents are in the tools directory.

I can't `dotnet workload search suggest` or `dotnet tool search maui` to find the same packages.

So, in fact this is prior-art of:
* Using a different package layout, and therefore not using `<PackAsTool>`
* Using a different package type, so search works differently.
* On install, the directory layout is different.

In other words, I see this as prior art that other teams writing more code to provide a more tailored experience than what a simple wrap of `dotnet tool` would provide. For their use case, the additional effort was worth the benefit. Therefore, I wonder if a small effort to make nuget plugins more flexible is worth the effort.